bindings/java/src/main/java/org/ucx/jucx/ucp/UcpWorkerParams.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (C) Mellanox Technologies Ltd. 2019. ALL RIGHTS RESERVED.
+ * See file LICENSE for terms.
+ */
+
+package org.ucx.jucx.ucp;
+
+import java.nio.ByteBuffer;
+import java.util.BitSet;
+
+import org.ucx.jucx.ucs.UcsConstatns;
+
+public class UcpWorkerParams {
+
+    /**
+     * Mask of valid fields in this structure.
+     * Fields not specified in this mask would be ignored.
+     * Provides ABI compatibility with respect to adding new fields.
+     */
+    private long fieldMask;
+
+
+    private int threadMode;
+
+    private BitSet cpuMask = new BitSet();
+
+    private long events;
+
+    private ByteBuffer userData;
+
+    private int eventFD;
+
+    /**
+     * Suggests the thread safety mode which workerand the associated resources
+     * should be created with. The default value is UCS_THREAD_MODE_SINGLE and
+     * it is used when the value of the parameter is not set. When this
+     * parameter is set, the {@link UcpWorker#UcpWorker(UcpContext)}
+     * attempts to create worker with this thread mode.
+     * The thread mode with which worker is created can differ from the
+     * suggested mode.
+     */
+    public UcpWorkerParams setThreadMode(UcsConstatns.UcsThreadMode threadMode) {
+        this.fieldMask |= UcpConstants.UCP_WORKER_PARAM_FIELD_THREAD_MODE;
+        this.threadMode = threadMode.ordinal();
+        return this;
+    }
+
+    /**
+     * Mask of which CPUs worker resources should preferably be allocated on.
+     * This value is optional.
+     * If it's not set, resources are allocated according to system's default policy.
+     */
+    public UcpWorkerParams setCpu(int cpuNum) {
```

Comment:
why setters are not having void sig ?
```
public void setCpuNum(int cpuNum)
{
 }
```

Reply:
It allows to chain all setters, like [here](https://github.com/openucx/ucx/pull/3256/files#diff-cb0fa86fe6141c99d49847756c138b76R67):
```
 UcpWorkerParams rdmaWorkerParams = new UcpWorkerParams()
                    .requestWakeupRMA().setCpu(i).setUserData(userData)
                    .setThreadMode(UcsConstatns.UcsThreadMode.UCS_THREAD_MODE_MULTI);
```
It's quite [typical patter ](https://dzone.com/articles/design-patterns-the-builder-pattern) in java.

Reply:
yep, it is typical cure to overcome telescopic constructors, which is not the case here. usually builder pattern contradicts OO principles and makes code ugly, hard to read and hard to set breakpoints.

Reply:
It's totally possible to use line by line:
```java
UcpWorkerParams rdmaWorkerParams = new UcpWorkerParams();
rdmaWorkerParams.requestWakeupRMA();
rdmaWorkerParams.setUserData(userData);
```
But on many options it makes code much longer. Even in core java it's videly used:
```java
ByteBuffer.allocateDirect(100).order(ByteOrder.BIG_ENDIAN).asCharBuffer().put("1").asReadOnlyBuffer()
```
In Netty (network library for java):
```java
ServerBootstrap b = new ServerBootstrap()
.option(ChannelOption.SO_BACKLOG, 128)
.childOption(ChannelOption.SO_KEEPALIVE, true);
```

And in many other places.

Reply:
Btw this example is better because it uniform api as setter style.
In pr there are many dedicated request funcs with diff sig.

Reply:
`setXXX` is to set primitive type field (like the number, etc.), while `requestXXX` is to request feature from enum. Calling `setXXX` second time will override previous value, while `requestFeatureXXX` will add bit flag to requested previously features. We discussed it here: https://github.com/openucx/ucx/pull/3214#pullrequestreview-203350761
To hide all these bit masks from a user. "Builder" pattern is used to quickly construct parameters object, that has no side effects. Side effects calls to JNI is done on construction `UcpContext` and `UcpWorker` objects.

Reply:
IMO builder pattern looks good for parameters object w/o side effects.